mobile/library/java/io/envoyproxy/envoymobile/engine/AndroidNetworkMonitorV2.java

```diff
@@ -0,0 +1,809 @@
+      case TelephonyManager.NETWORK_TYPE_HSPA:
+      case TelephonyManager.NETWORK_TYPE_EVDO_B:
+      case TelephonyManager.NETWORK_TYPE_EHRPD:
+      case TelephonyManager.NETWORK_TYPE_HSPAP:
+        return EnvoyConnectionType.CONNECTION_3G;
+      case TelephonyManager.NETWORK_TYPE_LTE:
+        return EnvoyConnectionType.CONNECTION_4G;
+      case TelephonyManager.NETWORK_TYPE_NR:
+        return EnvoyConnectionType.CONNECTION_5G;
+      default:
+        return EnvoyConnectionType.CONNECTION_UNKNOWN;
+      }
+    default:
+      return EnvoyConnectionType.CONNECTION_UNKNOWN;
+    }
+  }
+
+  public NetworkState(boolean connected, int type, int subtype, boolean isMetered,
+                      String networkIdentifier, boolean isPrivateDnsActive,
+                      String privateDnsServerName) {
+    mConnected = connected;
+    mType = type;
+    mSubtype = subtype;
+    mIsMetered = isMetered;
+    mNetworkIdentifier = networkIdentifier == null ? "" : networkIdentifier;
+    mIsPrivateDnsActive = isPrivateDnsActive;
+    mPrivateDnsServerName = privateDnsServerName == null ? "" : privateDnsServerName;
+  }
+
+  public boolean isConnected() { return mConnected; }
+
+  public int getNetworkType() { return mType; }
+
+  public boolean isMetered() { return mIsMetered; }
+
+  public int getNetworkSubType() { return mSubtype; }
+
+  // Always non-null to facilitate .equals().
+  public String getNetworkIdentifier() { return mNetworkIdentifier; }
+
+  /** Returns the connection type for the given NetworkState. */
+  public EnvoyConnectionType getEnvoyConnectionType() {
+    if (!isConnected()) {
+      return EnvoyConnectionType.CONNECTION_NONE;
+    }
+    return convertToEnvoyConnectionType(mType, mSubtype);
+  }
+
+  /** Returns boolean indicating if this network uses DNS-over-TLS. */
+  public boolean isPrivateDnsActive() { return mIsPrivateDnsActive; }
+
+  /** Returns the DNS-over-TLS server in use, if specified. */
+  public String getPrivateDnsServerName() { return mPrivateDnsServerName; }
+}
+
+/**
+ * This class does the following.
+ * <ul>
+ * <li>When the internet is available: call the
+ * <code>InternalEngine::onDefaultNetworkAvailable</code> callback.</li>
+ *
+ * <li>When the internet is not available: call the
+ * <code>InternalEngine::onDefaultNetworkUnavailable</code> callback.</li>
+ *
+ * <li>When the default network is changed: call the
+ * <code>EnvoyEngine::onDefaultNetworkChangedV2</code>.</li>
+ *
+ * <li>When any network is available: call the
+ * <code>EnvoyEngine::onNetworkConnect</code>.</li>
+ *
+ * <li>When any network is unavailable: call the
+ * <code>EnvoyEngine::onNetworkDisconnect</code>.</li>
+ *
+ * <li>When VPN network is available: call the
+ * <code>EnvoyEngine::purgeActiveNetworkList</code>.</li>
+ *
+ * <li>When VPN network is unavailable: call the
+ * <code>EnvoyEngine::onNetworkConnected</code> with all the rest of the connected networks.</li>
+ * </ul>
```

Comment:
Should we include an example of how this class should be used?  e.g.

```
AndroidNetworkMonitorV2 monitor = AndroidNetworkMonitorV2.load(...);
// I think you need to call this next?
monitor.registerNetworkCallbacks();
// What should be called afterwards?
```

Reply:
These are all wrapped into AndroidNetworkMonitorV2.load(). All we need to do is to swap in this class where AndroidNetworkMonitor.load() is called. Of course, behind some sort of runtime guard or Cronet API.

Reply:
Oh I see now, the private constructor does all the setup